doc/guides/naming_conventions/index.rst

```diff
@@ -0,0 +1,81 @@
+   ``protected``, and many others.
+
+#. Function names are always in lowercase, with an underscore between each part of
+   the name : ``k_timer_init()`` , ``k_timer_start()``.
+
+#. If a function operates primarily on a certain type, the prefix of this function
+   corresponds to the type name. For example, the k_timer_* functions work with the
+   ``k_timer`` type, and k_sem_* functions go with ``k_sem``.
+
+#. Each procedure’s name shall be descriptive of its purpose. Note that procedures
+   encapsulate the “actions” of a program and thus benefit from the use of verbs in
+   their names (e.g., ``k_mutex_lock()``); this “noun-verb” word ordering is recommended.
+   Alternatively, procedures may be named according to the question they answer
+   (e.g., ``device_is_ready()``).
+
+#. The names of all public functions shall be prefixed with their module name and
+   an underscore (e.g., ``sensor_read()``).
+
+#. Avoid names that differ only in case, like foo and Foo. Similarly, avoid foobar
+   and foo_bar. The potential for confusion is considerable.
+
+#. Similarly, avoid names that look like each other. On many terminals and
+   printers, 'l', '1' and 'I' look quite similar. A variable named 'l' is
+   particularly bad because it looks so much like the constant '1'.
+
+#. #define constants should be in all CAPS.
+
+#. No procedure shall have a name that overlaps a function in the C Standard
+   Library. Examples of such names include strlen, atoi, and memset.
+
+#. No procedure shall have a name that begins with an underscore.
+
+#. No procedure name shall be longer than 31 characters.
+
+#. Variable names, structure tag names, and function names should be in lower
+   case.
+
+#. A prototype shall be declared for each public function in the module header
+   file.
+
+#. All private functions shall be declared ``static``.
+
+#. Each parameter of a function shall be explicitly declared and meaningfully
+   named.
+
+#. Parameterized macros shall not be used if a function can be written to
+   accomplish the same behavior.
+
+
+Variables
+*********
+
+#. No variable shall have a name that is a keyword of C, C++, or any other
+   well-known extension of the C programming language, including specifically K&R C
+   and C99. Restricted names include ``interrupt``, ``inline``, ``restrict``, ``class``, ``true``,
+   ``false``, ``public``, ``private``,``friend``, and ``protected``.
+
+#. No variable shall have a name that overlaps with a variable name from the
+   C Standard Library (e.g., ``errno``).
+
+#. No variable name shall be longer than 31 characters.
+
+#. No variable name shall be shorter than 3 characters, including loop counters.
+
+#. No variable name shall contain any uppercase letters.
+
+#. Underscores shall be used to separate words in variable names.
+
+#. Each variable’s name shall be descriptive of its purpose.
+
+#. The names of any global variables shall begin with the letter ‘g’. For
+   example, ``g_zero_offset``.
```

Comment:
How is global defined? 

Is a global variable a static variable that is defined outside of any function? Are variables that are defined outside of any function but is not static considered the same level of global, even if it's exposed via a header file with `extern` like `extern struct bt_dev bt_dev;`? 

Similarly, should we require/recommend that any such external variables shall be declared static and exposed via an API instead?

Reply:
Let's assume global means "not automatic". Aside from this, I completely disagree with this rule, I have never believed in encoding the type or nature of a variable in the variable name itself.